File: src/mastermind.hs
import System.Random
import Data.List
import System.IO.Unsafe
import System.IO
import qualified Data.Text as T
import Text.Regex.Posix

-- These are actually functions
solutionLength = 4
maxGuesses = 10
guessWrong = "_"
guessRight = "X"
guessPresent = "*"
choices = "ABCDEF"

-- Library code
gameIntro :: IO ()
gameIntro = do
    -- Normally, " " (space) has the highest precendent, and so this executes fine
    putStrLn "Let's play Mastermind! The rules are easy, I promise."
    -- If we start doing other actions, we use $ to say "evaluate everything to the right first"
    -- Without this, the next line would be executed as: (putStrLn "I will pick ") ++ "something"
    -- With a $, it becomes then: putStrLn ("a" ++ "b")
    putStrLn $ "I will pick " ++ (show solutionLength) ++ " letters out of a possible " ++ (show (length choices)) ++ "."
    putStrLn $ "You will then get " ++ (show maxGuesses) ++ " chances to guess which letters I picked."
    putStrLn $ "For each letter you guess, I will answer as follows:"
    putStrLn $ "\t" ++ guessWrong ++ " means you got it completely wrong."
    putStrLn $ "\t" ++ guessPresent ++ " means you guessed a letter I used, but it's in the wrong position."
    putStrLn $ "\t" ++ guessRight ++ " means you guessed that letter right!"
    putStrLn ""


-- If real Haskell programmers see this function, I am pretty sure they will cry
randomStr :: (RandomGen g) => g -> [Char]
randomStr g = do
    take 1 $ randomRs ('A','F') $ unsafePerformIO newStdGen :: [Char]


-- <function name> :: <type constraints> => <signature>
generateSolutionWorker :: (Num a, Eq a) => a -> [Char] -> StdGen -> [Char]
generateSolutionWorker 0 acc@(h:t) g = acc
generateSolutionWorker n acc g
    | letter `isInfixOf` acc = generateSolutionWorker n acc g
    | otherwise = generateSolutionWorker (n - 1) (letter ++ acc) g
    where letter = (randomStr g)

generateSolution :: StdGen -> [Char]
generateSolution g = generateSolutionWorker solutionLength "" g


isValidGuess :: String -> Bool
isValidGuess guess
    | isTotalMatch = True
    | otherwise = False
    where pat = "[^" ++ (choices) ++ "]"
          isTotalMatch = isMatch == False && rightLength == True
          isMatch = (guess =~ pat) :: Bool
          rightLength = (length guess) == (solutionLength)


getGuess :: IO String
getGuess = do
    guess <- askGuess
    if (isValidGuess guess)
        then return guess
        else do
            putStrLn "We were unable to understand your input. Please enter only the letters of your guess and press enter."
            getGuess

askGuess :: IO String
askGuess = do
    let choiceList = [(head choices)] ++ (concat [", " ++ [x] | x <- (tail choices)])
    putStrLn ("Available pegs: " ++ choiceList)
    putStr ("Enter your guess (pick " ++ (show solutionLength) ++ "): ")
    hFlush stdout
    line <- getLine
    -- . joins functions T.unpack . T.toUpper $ <var> == (T.unpack (T.toUpper <var>))
    return $ T.unpack . T.toUpper . T.strip $ T.pack line


analyzeGuessWorker :: String -> String -> String -> String -> String
analyzeGuessWorker _ [] _ acc = reverse acc
analyzeGuessWorker _ _ [] acc = reverse acc
analyzeGuessWorker solution (solutionLetter : solutionRest) (guessLetter : guessRest) acc
    | isSame = analyzeGuessWorker solution solutionRest guessRest (guessRight ++ acc)
    | isInSolution = analyzeGuessWorker solution solutionRest guessRest (guessPresent ++ acc)
    | otherwise = analyzeGuessWorker solution solutionRest guessRest (guessWrong ++ acc)
    where isSame = guessLetter == solutionLetter
          isInSolution = [guessLetter] `isInfixOf` solution


analyzeGuess :: String -> String -> String
analyzeGuess [] _ = []
analyzeGuess _ [] = []
analyzeGuess solution guess = analyzeGuessWorker solution solution guess []

-- Game code
startGame :: StdGen -> IO ()
startGame g = do
    gameIntro
    let solution = (generateSolution g)
    gameLoop solution maxGuesses


gameLoop :: String -> Integer -> IO ()
gameLoop solution 0 = putStrLn "You've ran out of guesses! Better luck next time!"
gameLoop solution iteration = do
    guess <- getGuess
    if guess == solution
        then putStrLn "You've guessed correctly! You win!"
        else do
            putStrLn ("Result: " ++ (analyzeGuess solution guess))
            gameLoop solution (iteration - 1)


main:: IO ()
main = do
    g <- newStdGen
    startGame g

File: src/mastermind.cs
using System;
using System.Text.RegularExpressions;
 
public class Mastermind
{
	static public int SOLUTION_LENGTH = 4;
	static public int MAX_GUESSES = 10;
	static public char GUESS_RIGHT = 'X';
	static public char GUESS_PRESENT = '*';
	static public char GUESS_WRONG = '_';
	static public string CHOICES = "ABCDEF";

	public void GameIntro() {
	    Console.WriteLine("Let's play Mastermind! The rules are easy, I promise.");
	    Console.WriteLine("I will pick " + SOLUTION_LENGTH + " letters out of a possible " + CHOICES.Length + ".");
	    Console.WriteLine("You will then get " + MAX_GUESSES + " chances to guess which letters I picked.");
	    Console.WriteLine("For each letter you guess, I will answer as follows:");
	    Console.WriteLine("\t" + GUESS_WRONG + " means you got it completely wrong.");
	    Console.WriteLine("\t" + GUESS_PRESENT + " means you guessed a letter I used, but it's in the wrong position.");
	    Console.WriteLine("\t" + GUESS_RIGHT + " means you guessed that letter right!");
	    Console.WriteLine("");

	}

	public string GenerateSolution() {
		Random rnd = new Random();
		string solution = "";
		while (solution.Length < SOLUTION_LENGTH) {
			int pick = rnd.Next(CHOICES.Length);
			char choice = CHOICES[pick];
			if (solution.IndexOf(choice) == -1) {
				solution = solution + choice;
			}
		}

		return solution;
	}

	public string AskGuess() {
		Console.WriteLine("Available pegs: " + CHOICES);
		Console.Write("Enter your guess (pick " + SOLUTION_LENGTH + "): ");
		string guess = Console.ReadLine();
		return guess.Trim().ToUpper();
	}

	public bool IsGuessValid(string guess) {
		if (guess.Length != SOLUTION_LENGTH) {
			return false;
		}

		string pattern = "[^" + CHOICES + "]";

		if (Regex.IsMatch(guess, pattern)) {
			return false;
		}

		return true;
	}

	public string GetGuess() {
		string guess = AskGuess();
		while (IsGuessValid(guess) == false) {
			Console.WriteLine("We were unable to understand your input. Please enter only the letters of your guess and press enter.");
			guess = AskGuess();
		}

		return guess;
	}

	public string AnalyzeGuess(string solution, string guess) {
		string result = "";

		for (int i = 0; i < SOLUTION_LENGTH; i++) {
			char solution_letter = solution[i];
			char guess_letter = guess[i];

			if (solution_letter == guess_letter) {
				result = result + GUESS_RIGHT;
			} else if (solution.IndexOf(guess_letter) > -1) {
				result = result + GUESS_PRESENT;
			} else {
				result = result + GUESS_WRONG;
			}
		}

		return result;
	}

	public void StartGame() {
		GameIntro();

		string solution = GenerateSolution();
		int guesses = 0;

		while(true) {
			if (guesses > MAX_GUESSES) {
				Console.WriteLine("You've ran out of guesses! Better luck next time!");
				return;
			}

			string guess = GetGuess();

			if (guess == solution) {
				Console.WriteLine("You've guessed correctly, you win!");
				return;
			}

			Console.WriteLine("Result: " + AnalyzeGuess(solution, guess));

			guesses = guesses + 1;
		}
	}

	static public void Main ()
	{
		Mastermind game = new Mastermind();
		game.StartGame();
	}
 
}
